test/tests_for_integration/api_to_db_test.py

```diff
@@ -607,3 +617,149 @@ def _run_job_fail(ee2_port, token, params, expected, throw_exception=False):
         with raises(ServerError) as got:
             client.run_job(params)
         assert_exception_correct(got.value, ServerError("name", 1, expected))
+
+
+######## run_job_concierge tests ########
+
+
+def test_run_job_concierge_minimal(ee2_port, ws_controller, mongo_client):
+    def modify_sub(sub):
+        del sub["Concurrency_Limits"]
+
+    _run_job_concierge(
+        ee2_port,
+        ws_controller,
+        mongo_client,
+        # if the concierge dict is empty, regular old run_job gets run
+        {"trigger": "concierge"},  # contents are ignored
+        modify_sub,
+        "concierge",
+        4,
+        23000,
+        100,
+    )
+
+
+def test_run_job_concierge_mixed(ee2_port, ws_controller, mongo_client):
+    """
+    Gets cpu from the input, memory from deploy.cfg, and disk from the catalog.
+    """
+
+    def modify_sub(sub):
+        del sub["Concurrency_Limits"]
+
+    _run_job_concierge(
+        ee2_port,
+        ws_controller,
+        mongo_client,
+        {"client_group": "extreme", "request_cpus": 76},
+        modify_sub,
+        "extreme",
+        76,
+        250000,
+        7,
+        [{"client_groups": ['{"request_cpus":8,"request_disk":7}']}],
+    )
+
+
+def test_run_job_concierge_maximal(ee2_port, ws_controller, mongo_client):
+    def modify_sub(sub):
+        sub[
+            "requirements"
+        ] = '(CLIENTGROUP == "bigmem") && (baz == "bat") && (foo == "bar")'
+        sub["Concurrency_Limits"] = "some_sucker"
+        sub["+AccountingGroup"] = '"some_sucker"'
+        sub["environment"] = sub["environment"].replace(
+            "DEBUG_MODE=False", "DEBUG_MODE=True"
+        )
+
+    _run_job_concierge(
+        ee2_port,
+        ws_controller,
+        mongo_client,
+        {
+            "client_group": "bigmem",
+            "request_cpus": 42,
+            "request_memory": 56,
+            "request_disk": 89,
+            "client_group_regex": False,
+            "account_group": "some_sucker",
+            "ignore_concurrency_limits": False,
+            "requirements_list": ["foo=bar", "baz=bat"],
+            "debug_mode": "true",
+        },
+        modify_sub,
+        "bigmem",
+        42,
+        56,
+        89,
+    )
+
+
+def _run_job_concierge(
+    ee2_port,
+    ws_controller,
+    mongo_client,
+    conc_params,
+    modify_sub,
+    clientgroup,
+    cpu,
+    mem,
+    disk,
+    catalog_return=None,
+):
+    _set_up_workspace_objects(ws_controller, TOKEN_KBASE_CONCIERGE)
+    # need to get the mock objects first so spec_set can do its magic before we mock out
+    # the classes in the context manager
+    sub, schedd, txn = _get_htc_mocks()
+    # seriously black you're killing me here. This is readable?
```

Comment:
I don't think black is necessarily optimised for readability. It's just for consistency.

Reply:
Agreed. I have no issues if you would like to adjust black formatting params, just as long as the entire codebase is formatted one way.

Reply:
To me readability is a lot more important than consistency

Reply:
I thought black basically didn't have adjustable parameters - it was modeled after gofmt, wasn't it?

Reply:
And honestly I'm just whinging. It's not worth spending time on messing with black, even if that's possible, for this one case.

Reply:
It's nice to have a bit of a whinge once in a while.

You should see the things that prettier does to JS code! It is an abomination.

Reply:
~Sometimes~ I think linters aren't worth the trouble.